core/math/a_star_grid_2d.cpp

```diff
@@ -734,6 +734,121 @@ TypedArray<Vector2i> AStarGrid2D::get_id_path(const Vector2i &p_from_id, const V
+}
+
+void AStarGrid2D::set_point_cost(const Vector2i &p_id, int p_cost) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_MSG(!is_in_boundsv(p_id), vformat("Can't set point's cost. Point %s out of bounds %s.", p_id, region));
+	ERR_FAIL_COND_MSG(p_cost < 1 || p_cost > 254, "Can't set point's cost. Cost must be within (1-254) range.");
+	_get_point_unchecked(p_id)->cost = p_cost;
+}
+
+int AStarGrid2D::get_point_cost(const Vector2i &p_id) const {
+	ERR_FAIL_COND_V_MSG(dirty, 0, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_V_MSG(!is_in_boundsv(p_id), 0, vformat("Can't get point's cost. Point %s out of bounds %s.", p_id, region));
+	return _get_point_unchecked(p_id)->cost;
+}
+
+void AStarGrid2D::fill_flow_field_region(const Rect2i &p_region, const Vector2 &p_value) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+
+	const Rect2i safe_region = p_region.intersection(region);
+	const int32_t end_x = safe_region.get_end().x;
+	const int32_t end_y = safe_region.get_end().y;
+
+	for (int32_t y = safe_region.position.y; y < end_y; y++) {
+		for (int32_t x = safe_region.position.x; x < end_x; x++) {
+			_get_point_unchecked(x, y)->flow_direction = p_value;
+		}
+	}
+}
+
+void AStarGrid2D::calculate_flow_field(const Vector2i &p_id, const Rect2i &p_region, bool p_clear_previous_region) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+
+	Rect2i region_to_check;
+	if (p_region == Rect2i()) {
+		region_to_check = region;
+	} else {
+		region_to_check = p_region.intersection(region);
+	}
+
+	ERR_FAIL_COND_MSG(!region_to_check.has_point(p_id), vformat("Can't calculate flow field. Point %s out of bounds %s.", p_id, region_to_check));
+
+	if (p_clear_previous_region) {
+		const int32_t end_x = previous_ff_region.get_end().x;
+		const int32_t end_y = previous_ff_region.get_end().y;
+
+		for (int y = previous_ff_region.position.y; y < end_y; y++) {
+			for (int x = previous_ff_region.position.x; x < end_x; x++) {
+				_get_point_unchecked(x, y)->flow_direction = Vector2();
+			}
+		}
+	}
+	previous_ff_region = region_to_check;
+
+	LocalVector<uint32_t> values;
+	values.resize(region_to_check.size.x * region_to_check.size.y);
+
+	{
+		size_t index = 0U;
+		uint32_t *ptr = values.ptr();
+
+		const int32_t end_x = region_to_check.get_end().x;
+		const int32_t end_y = region_to_check.get_end().y;
+
+		for (int y = region_to_check.position.y; y < end_y; y++) {
+			for (int x = region_to_check.position.x; x < end_x; x++) {
+				ptr[index++] = UINT32_MAX;
+			}
+		}
+	}
+
+	LocalVector<Point *> nbors;
+	nbors.reserve(16);
+
+	LocalVector<Point *> open_list;
+
+	Point *target = _get_point(p_id.x, p_id.y);
+	values[_to_index(p_id, region_to_check)] = 0;
+	open_list.push_back(target);
+
+	while (!open_list.is_empty()) {
+		Point *p = open_list[open_list.size() - 1];
+		open_list.remove_at(open_list.size() - 1);
```

Comment:
I'd say `open_list` should be a priority queue to ensure proper iteration order (by non-descending cost). Otherwise the same point could be enlisted many times which theoretically should result in worse time complexity (probably needs benchmarking though :thinking:).

There's this max heap implementation:
https://github.com/godotengine/godot/blob/6dc78c8aa17ad46e701e0c4e7b7632c2f0e098f1/modules/navigation/nav_utils.h#L193-L311

Min heap would do here so the above could be used with different `LessThan` comparator (`Indexer` is not needed here though).